src/serializer/ResidualHeapVisitor.js

```diff
@@ -695,10 +729,53 @@ export class ResidualHeapVisitor {
       });
       this.visitGenerator(generator);
       this._withScope(generator, visitPropertiesAndBindings);
+
+      // Remove any modifications to CreatedObjects -- these are fine being serialized inside the additional function
+      this.reVisitSet = new Set([...this.reVisitSet].filter(x => !createdObjects.has(x)));
+
       this.realm.restoreBindings(modifiedBindings);
       this.realm.restoreProperties(modifiedProperties);
-      this.reactElementEquivalenceSet = oldReactElementEquivalenceSet;
+      return this.realm.intrinsics.undefined;
+    };
+    this.realm.evaluateAndRevertInGlobalEnv(_visitAdditionalFunctionEffects);
+
+    // Cleanup
+    this.commonScope = prevCommonScope;
+    this.reactElementEquivalenceSet = oldReactElementEquivalenceSet;
+    this._withScope(
+      parentScope,
+      // Re-visit any bindings corresponding to unbound values or values closed over from outside additional function
+      // they're serialized in the correct scope
+      () => {
+        invariant(functionInfo !== undefined);
+        invariant(funcInstance !== undefined);
+        for (let value of this.reVisitSet) {
+          // Populate old reVisitSet because we switched them out
+          prevReVisit.add(value);
+          this.visitValue(value);
+        }
+        for (let innerName of functionInfo.unbound) {
+          let residualBinding = this.visitBinding(functionValue, innerName, false);
+          if (residualBinding) {
+            residualBinding.modified = true;
+            funcInstance.residualFunctionBindings.set(innerName, residualBinding);
+          }
+        }
+        this.reVisitSet = prevReVisit;
+      }
+    );
+    this.inAdditionalFunction = false;
```

Comment:
This subtly assumes that additional functions can never be nested inside additional functions.